pkg/storage/ingest/reader.go

```diff
@@ -1122,20 +1144,50 @@ func handleKafkaFetchErr(fr fetchResult, fw fetchWant, shortBackoff, longBackoff
 	case errors.Is(err, kerr.NotLeaderForPartition):
-		longBackoff.Wait() // our metadata is out of date, we should refresh and try again with a leader who is up to date // TODO this current doesn't happen
+		// We're asking a broker which is no longer the leader. For a partition. We should refresh our metadata and try again.
+		triggerMetadataRefresh(refresher)
+		longBackoff.Wait()
 	case errors.Is(err, kerr.ReplicaNotAvailable):
-		longBackoff.Wait() // our metadata is out of date, we should refresh and try again with a leader who is up to date // TODO this current doesn't happen
+		// Maybe the replica hasn't replicated the log yet, or it is no longer a replica for this partition.
+		// We should refresh and try again with a leader or replica which is up to date.
+		triggerMetadataRefresh(refresher)
+		longBackoff.Wait()
 	case errors.Is(err, kerr.UnknownLeaderEpoch):
-		longBackoff.Wait() // our metadata is out of date, we should refresh and try again with a leader who is up to date // TODO this current doesn't happen
+		// Maybe there's an ongoing election. We should refresh our metadata and try again with a leader in the current epoch.
+		triggerMetadataRefresh(refresher)
+		longBackoff.Wait()
 	case errors.Is(err, kerr.FencedLeaderEpoch):
-		longBackoff.Wait() // our metadata is out of date, we should refresh and try again with a leader who is up to date // TODO this current doesn't happen
+		// We missed a new epoch (leader election). We should refresh our metadata and try again with a leader in the current epoch.
+		triggerMetadataRefresh(refresher)
+		longBackoff.Wait()
+	case errors.Is(err, kerr.LeaderNotAvailable):
+		// This isn't listed in the possible errors in franz-go, but Apache Kafka returns it when the partition has no leader.
+		triggerMetadataRefresh(refresher)
+		longBackoff.Wait()
+	case errors.Is(err, errUnknownPartitionLeader):
+		triggerMetadataRefresh(refresher)
+		longBackoff.Wait()
+	case errors.Is(err, &kgo.ErrFirstReadEOF{}):
+		longBackoff.Wait()
+
 	default:
 		level.Error(logger).Log("msg", "received an error we're not prepared to handle; this shouldn't happen; please report this as a bug", "err", err)
 		longBackoff.Wait()
 	}
 	return fw
 }
 
+func triggerMetadataRefresh(client metadataRefresher) {
+	// Typically franz-go will update its own metadata when it detects a change in brokers. But it's hard to verify this.
+	// So we force a metadata refresh here to be sure.
+	// It's ok to call this from multiple fetchers concurrently. franz-go will only be sending one metadata request at a time (whether automatic, periodic, or forced).
+	//
+	// Metadata refresh is asynchronous. So even after forcing the refresh we might have outdated metadata.
+	// Hopefully the backoff that will follow is enough to get the latest metadata.
+	// If not, the fetcher will end up here again on the next attempt.
+	client.ForceMetadataRefresh()
+}
+
```

Comment:
nit: I don't think we need this level of interaction - if you care more about the comment than anything else, you can just inline the metadata refresh in a closure within `handleKafkaFetchErr`.

e.g.
```golang
func handleKafkaFetchErr(fr fetchResult, fw fetchWant, shortBackoff, longBackoff waiter, client *kgo.Client, logger log.Logger) fetchWant {
	refreshMetadata := client.ForceMetadataRefresh
```

and then add your comment on top, I do understand that this makes it simpler to test but for now, It feels like we should have a `ReaderClient` or something interface around here, which is kind of a different problem.

Reply:
i do care about the comment and can inline this function, but what's wrong with the interface? we have the two `waiter`s too? for now we need only this if we later have a need for `ReaderClient` we can replace `metadataRefresher`, but we don't now (do we?)

Reply:
I was more concerned with the indirection than interface to be honest. I'm generally not a fan of single method interfaces but can appreciate it's very subjective, this LGTM.